src/java/org/apache/cassandra/tools/nodetool/stats/DataPathsHolder.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.tools.nodetool.stats;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.cassandra.db.ColumnFamilyStoreMBean;
+import org.apache.cassandra.tools.NodeProbe;
+
+public class DataPathsHolder implements StatsHolder
+{
+    public final Map<String, Object> pathsHash;
+
+    public DataPathsHolder(NodeProbe probe, List<String> tableNames)
+    {
+        this.pathsHash = new HashMap();
+
+        Iterator<Map.Entry<String, ColumnFamilyStoreMBean>> tableMBeans = probe.getColumnFamilyStoreMBeanProxies();
+        while (tableMBeans.hasNext())
+        {
+            Map.Entry<String, ColumnFamilyStoreMBean> entry = tableMBeans.next();
+            String keyspaceName = entry.getKey();
+            String tableName = entry.getValue().getTableName();
+
+            if (!(tableNames.isEmpty() || 
+                  tableNames.contains(keyspaceName + "." + tableName) || 
+                  tableNames.contains(keyspaceName) ))
+            {
+                continue;
+            }
+
+            Map<String, List<String>> ksPaths;
+            List<String> dataPaths;
+
+            try
+            {
+                dataPaths = entry.getValue().getDataPaths();
+            }
+            catch (IOException ex)
+            {
+                continue;
```

Comment:
This should be logged with logger on warn level.

Reply:
Does nodetool has a logger? Or just print a warning to stderr?

Reply:
I was looking into it, and I don't understand why [ColumnFamilyStoreMBean.getDataPaths()](https://github.com/apache/cassandra/blob/d80bb323c9dd8f814fe02e16dec510557c7d5101/src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java#L268) does announce IOException to be thrown. The only productive code implementing that interface is [ColumnFamilyStore.getDataPaths()](https://github.com/apache/cassandra/blob/d80bb323c9dd8f814fe02e16dec510557c7d5101/src/java/org/apache/cassandra/db/ColumnFamilyStore.java#L600) which doesn't seem to throw any exception. Haven't check the bunch of MockCFS implementations, but it seems they are for testing purposes only. Did I miss something?

Reply:
If you checkout the original commit which was pushed, back in time it was possible that the exception might be thrown with `dataPath.getCanonicalPath()`

But then there was patch of CASSANDRA-16926 CEP-10 Phase 1, which changed that code so it stopped to throw anything.

If you dig deep enough, there is `PathUtils.toRealPath` called on that line which wrapps IOException and it throws unchecked exception and if you eventually dig even further it will ultimately wrap that IOError into FSReadError which is runtime exception.

Hence, I would change this to catching `Throwable t` there and I would log that one.

https://issues.apache.org/jira/browse/CASSANDRA-16335

Reply:
You can reuse

        output.err.println("error: " + e.getMessage());
        output.err.println("-- StackTrace --");
        output.err.println(getStackTraceAsString(e));

This code is in `NodeTool.err(Throwable)` method but it is protected so it can not be called directly in `DataPathsHolder` but since you have access to `NodeProbe` in that method, you can call it via `probe.output().err.println()`.

Reply:
@rtib please `catch (Throwable t)` instead of IOException

Reply:
Changed that, but some IDE (e.g. NetBeans) hint on that to change back.

Reply:
Yes because it still thinks that we are throwing that from the method based on its signature which is, as explained not true. To be safe, lets just catch it all and thats it.